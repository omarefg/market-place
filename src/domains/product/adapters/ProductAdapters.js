import React, { createContext, useState } from "react";

export const ProductContext = createContext();

const loadersKeys = {
  getAll: "getAll",
};

const errorsKeys = {
  getAll: "getAll",
};

export function ProductProvider(props) {
  const { components, children, useCases, handlersBuilder } = props;

  const [products, setProducts] = useState([]);
  const [loaders, setLoaders] = useState({});
  const [errors, setErrors] = useState({});

  const value = {
    state: {
      products,
      loaders,
      errors,
    },
    handlers: handlersBuilder({
      stateSetters: {
        setProducts,
        setLoaders,
        setErrors,
      },
      loadersKeys,
      errorsKeys,
      useCases
    }),
    components,
    metadata: {
      loadersKeys,
      errorsKeys,
    },
  };


  return (
    <ProductContext.Provider value={value}>{children}</ProductContext.Provider>
  );
}
