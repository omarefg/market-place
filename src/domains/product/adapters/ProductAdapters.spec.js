import React from "react";
import { render } from "@testing-library/react";
import { ProductProvider } from "./ProductAdapters";

describe("@domains/product/adapters/ProductProvider", () => {
  let defaultProps, wrapper;

  beforeEach(() => {
    defaultProps = {
      components: {},
      children: <div data-testid="ProductAdapter__children" />,
      useCases: {},
      handlersBuilder: jest.fn(),
    };

    wrapper = (props = {}) => {
      const finalProps = { ...defaultProps, ...props };
      return render(<ProductProvider {...finalProps} />);
    };
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it("Given a valid call it must render withot crashing", () => {
    // Arrange
    const { container } = wrapper();
    // Assert
    expect(container).toMatchSnapshot();
  });

  it("Given a valid call it must render one child node", () => {
    // Arrange
    const { container } = wrapper();
    // Act
    const hasOneChildNode = container.childNodes.length === 1;
    // Assert
    expect(hasOneChildNode).toBe(true);
  });

  it("Given a valid call it must render `children` prop", () => {
    // Arrange
    const { container, getByTestId } = wrapper();
    // Act
    const children = getByTestId("ProductAdapter__children");
    //Assert
    expect(container).toContainElement(children);
  });

  it("Given a valid call it must call handlersBuilder to generate builders in value", () => {
    // Act
    wrapper();
    //Assert
    expect(defaultProps.handlersBuilder).toBeCalledTimes(1);
  });
});
