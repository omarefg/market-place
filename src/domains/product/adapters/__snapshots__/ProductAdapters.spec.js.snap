// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@domains/product/adapters/ProductProvider Given a valid call it must render withot crashing 1`] = `
<div>
  <div
    data-testid="ProductAdapter__children"
  />
</div>
`;
